### ISSUE TITLE
Degraded md raid being inactive while using with lvm

### ISSUE THREAD

**Issue by kcchuang** [dracutdevs/dracut#227]
Hi,

I'm using a /dev/md0 as my /boot partition and lvm for other partition, hence I add rd.md.uuid and rd.lvm.lv in grub.conf kernel parameter. When I plug out one of the block devices belonging to this /dev/md0, the /dev/md0 being inactive in initramfs and boot failed(enter user intervention mode). 
I look into dracut and find something interesting below, it will be appreciated if you could help ^_^.
I checked the latest master and the same logic as below, hence I think it's not thing about dracut version.
I upload the messages when rd.debug is open, https://uploadfiles.io/n4hjb.

**Environment:**
OS: centos7.2
dracut version: 033
kernel parameter: root=/dev/mapperr/os-root rd.lvm.lv=os/root rd.lvm.lv=os/swap rd.md.uuid=5068eccc:e5b886ea8:be8fe496:61ee1fd8 
output of lvs:
```
  LV   VG Attr       LSize  Pool Origin Data%  Meta%  Move Log Cpy%Sync Convert
  root os -wi-ao---- 75.00g
  swap os -wi-ao----  4.00g
```
output of 'mdadm -D /dev/md0':
```
     Raid Level : raid1
   Raid Devices : 3
  Total Devices : 3
          State : clean
 Active Devices : 3
Working Devices : 3
    Number   Major   Minor   RaidDevice State
       2       8        3        0      active sync   /dev/sda3
       4       8       19        1      active sync   /dev/sdb3
       3       8       48        2      active sync   /dev/sdd

```
**The code trace:**
1. /etc/udev/rules.d/65-md-incremental-imsm.rules:
`RUN+="/sbin/initqueue --timeout --name 50-mdraid_start --onetime --unique /sbin/mdraid_start"
`
Hence /sbin/initqueue will add a file '/lib/dracut/hooks/initqueue/timeout/50-mdraid_start.sh'

2. /etc/udev/rules.d/64-lvm.rules:

`RUN+="/sbin/initqueue --settled --onetime --unique /sbin/lvm_scan"`
`RUN+="/sbin/initqueue --timeout --name 51-lvm_scan --onetime --unique /sbin/lvm_scan --partial"`

Hence /sbin/initqueue will add 2 files  '/lib/dracut/hooks/initqueue/settled/lvm_scan.sh' and  '/lib/dracut/hooks/initqueue/timeout/51-lvm_scan.sh'

3. dracut-cmdline runs /lib/dracut/hooks/cmdline/30-parse-lvm.sh:
`wait_for_dev -n /dev/os/root`

Hence we have a file '/lib/dracut/hooks/initqueue/finished/devexists-\x2fdev\x2fmapper\x2fos-root.sh':
`[ -e /dev/mapper/os-root ]`

4. when systemd call /usr/bin/dracut-initqueue and dracut-initqueue start enter a while loop and check device status:

```
 24 while :; do

 30     check_finished && break

 44     for job in $hookdir/initqueue/settled/*.sh; do
 45         [ -e "$job" ] || break
 46         job=$job . $job
 47         check_finished && break 2
 48     done

 55     if [ $main_loop -gt $((2*$RDRETRY/3)) ]; then
 56         for job in $hookdir/initqueue/timeout/*.sh; do
 57             [ -e "$job" ] || break
 58             job=$job . $job
 59             udevadm settle --timeout=0 >/dev/null 2>&1 || main_loop=0
 60             [ -f $hookdir/initqueue/work ] && main_loop=0
 61         done
 62     fi
```

At line 30, the 'check_finished && break' won't break, because /dev/mapper/os-root doesn't exist yet.
At line 44, the first(only) job of the for loop is '/lib/dracut/hooks/initqueue/settled/lvm_scan.sh'
Hence after sourcing '/lib/dracut/hooks/initqueue/settled/lvm_scan.sh' the /dev/mapper/os-root exists.
At line 47, 'check_finished && break 2' will break the while loop and **there's no change to run line55 for the job in '/lib/dracut/hooks/initqueue/timeout/50-mdraid_start.sh.**

Is it a problem?

5. For now, I can do a workaround in /etc/udev/rules.d/65-md-incremental-imsm.rules:
`RUN+="/sbin/initqueue --finished --name 50-mdraid_start --onetime --unique /sbin/mdraid_start"
`
Thanks for any consideration.

**Comment by haraldh** [comment]
From the log file:

1. "console=ttyS0,9600": please use a higher baud rate!
1. "Starting Switch Root...": seems like the initramfs successfully booted
1. "Dependency failed for /boot.": seems like your /etc/fstab has a wrong entry for /boot

So, in this log file the problem is the missing /boot

**Comment by kcchuang** [comment]
Hi Harald Hoyer, thanks for your response ^^. Please check out my response below.

1. I uploaded another log with baud rate 115200 named typescript2 in https://uploadfiles.io/zb8gc. 
In this experiment, I press password to enter maintenance mode and execute "mdadm -D /dev/md0" and "blkid /dev/md0". I hope these info helps.
2. Yes, "Starting Switch Root...": seems like the initramfs successfully booted.
From my observation, /dev/md0 is still inactive even initramfs successfully booted, there's no any .sh in /lib/dracut/hooks/initqueue/finished/ to wait a degraded/inactive md becoming active.
3. My /etc/fstab looks like:
`UUID=21df6a4c-bce3-4fe4-8739-8e804f4893dd /boot           ext2  defaults        1 2`
21df6a4c-bce3-4fe4-8739-8e804f4893dd is blkid of /dev/md0. /dev/md0 is inactive, hence 21df6a4c-bce3-4fe4-8739-8e804f4893dd don't exist. I guess that why "Dependency failed for /boot."

Hence the problem is still /dev/md0 being inactive in initramfs.
Thanks for your help again.

**Comment by kcchuang** [comment]
Actually there's nothing to do with LVM, the problem also happen when there's no LVM. Sorry for my problem description makes you ambiguous.

**Comment by haraldh** [comment]
Ah.. Now I get your problem. Although /dev/md0 is not needed to boot to the real root, you want the initramfs to assemble it (even degraded) via "rd.md.uuid=5068eccc:e5b886ea8:be8fe496:61ee1fd8"

Right?

**Comment by haraldh** [comment]
The correct solution for this problem would be a initqueue/finished hook script, which checks for the existence of a md raid with the specified uuid.
This would lead to a timeout situation, which would trigger the initqueue/timeout scripts to run and which would activate the degraded raids.

**Comment by kcchuang** [comment]
Yes, you are right. 
I checked the history, there was a initqueue/finished/md_finished.sh hook script, but it was removed in commit https://github.com/dracutdevs/dracut/commit/a025cc17f0d8145492ffbb37735deca208e768bd.
I look and test the previous initqueue/finished/md_finished.sh hook script, it works.

**Comment by haraldh** [comment]
please check if commit 3cea065819d0a7ae9a32e097589510df315090d3 fixes your issue

**Comment by kcchuang** [comment]
Thank you very much, it fix my issue.

**Comment by kcchuang** [comment]
Hi Harald Hoyer, 

Thanks again for your hard effort.
May I know when I could get the latest dracut rpm including above fix item from centos update repo please?

**Comment by haraldh** [comment]
Before it hits "CentOS", it would have to hit Red Hat Enterprise Linux 7. Therefore we need an official bug entry on http://bugzilla.redhat.com . Then this bug has to be rated important enough for a backport. So, may I kindly ask you to file a bug describing your problem and reference this github issue?
Thanks!

**Comment by kcchuang** [comment]
Yes, I got you. I am trying to file an official bug on  http://bugzilla.redhat.com .
Thanks.

**Comment by shodanshok** [comment]
The problem described in this issue is not corrected in current RHEL 7.6. Moreover, it seems to be related to a too-high `rd.retry` timeout (180s), which conflict with the lower systemctl default timeout (90s). I can confirm that passing `rd.retry=30` to the boot command line solves the problem, enabling the array to start (albeit degraded) and system to boot. Finally, `dracut.cmdline` man page explicitly states that default `rd.retry` timeout should be 30s, rather than current 180s as found in `dracut-initqueue`

Please see here for more details: https://bugzilla.redhat.com/show_bug.cgi?id=1701504
